File: BE/Core/BookStore.Domain/Entities/Ordering & Payment/Order.cs
using BookStore.Domain.Entities.Identity;
using BookStore.Domain.Entities.Ordering___Payment;
using BookStore.Domain.Entities.Pricing___Inventory;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace BookStore.Domain.Entities.Ordering
{
    public class Order
    {
        public Guid Id { get; set; }

        public Guid UserId { get; set; }                       // Người mua hàng
        public virtual User User { get; set; } = null!;

        public string Status { get; set; } = "Pending";        // Trạng thái hiện tại (Pending, Paid, Shipped, Completed, Cancelled)
        public string OrderNumber { get; set; } = null!;       // Mã đơn hàng hiển thị cho khách (VD: BK-2025-0001)
        public decimal TotalAmount { get; set; }               // Tổng tiền đơn hàng
        public decimal DiscountAmount { get; set; }            // Tổng tiền giảm giá (từ Discount/Coupon)
        public decimal FinalAmount => TotalAmount - DiscountAmount; // Số tiền khách phải trả sau giảm

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        public DateTime? PaidAt { get; set; }                  // Thời điểm thanh toán (nếu đã trả)
        public DateTime? CompletedAt { get; set; }             // Khi đơn hoàn thành
        public DateTime? CancelledAt { get; set; }             // Khi đơn bị hủy

        // 🔗 1-n: Chi tiết sản phẩm
        public virtual ICollection<OrderItem> Items { get; set; } = new List<OrderItem>();

        // 🔗 1-1: Địa chỉ giao hàng riêng cho đơn
        public Guid AddressId { get; set; }
        public virtual OrderAddress Address { get; set; } = null!;

        // 🔗 1-1: Giao dịch thanh toán
        public virtual PaymentTransaction? PaymentTransaction { get; set; }

        // 🔗 1-n: Lịch sử trạng thái
        public virtual ICollection<OrderStatusLog> StatusLogs { get; set; } = new List<OrderStatusLog>();

        // 🔗 1-n: Lịch sử hành động chi tiết
        public virtual ICollection<OrderHistory> Histories { get; set; } = new List<OrderHistory>();

        // 🔗 n-1: Có thể dùng coupon
        public Guid? CouponId { get; set; }
        public virtual Coupon? Coupon { get; set; }
    }
}

File: web-frontend/src/types/order.ts
// Order Types
import type { Book } from './book';

export interface Order {
  id: string;
  userId: string;
  user?: User;
  status: OrderStatus;
  orderNumber: string;
  totalAmount: number;
  discountAmount: number;
  finalAmount: number;
  createdAt: string;
  paidAt?: string;
  completedAt?: string;
  cancelledAt?: string;

  // Relations
  items: OrderItem[];
  address: OrderAddress;
  paymentTransaction?: PaymentTransaction;
  couponId?: string;
  coupon?: Coupon;
}

export type OrderStatus =
  | "PendingPayment"
  | "Paid"
  | "Processing"
  | "Shipped"
  | "Completed"
  | "Cancelled"
  | "Refunded";

export interface OrderItem {
  id: string;
  orderId: string;
  bookId: string;
  book?: Book;
  quantity: number;
  unitPrice: number;
  totalPrice: number;
}

export interface OrderAddress {
  id: string;
  orderId: string;
  recipientName: string;
  phoneNumber: string;
  address: string;
  ward: string;
  district: string;
  province: string;
  postalCode?: string;
}

export interface PaymentTransaction {
  id: string;
  orderId: string;
  amount: number;
  status: PaymentStatus;
  transactionCode: string;
  paymentMethod: string;
  qrCodeUrl?: string;
  requestedAt: string;
  processedAt?: string;
}

export type PaymentStatus = "Pending" | "Paid" | "Failed" | "Cancelled";

export interface Coupon {
  id: string;
  code: string;
  discountValue: number;
  discountType: "Percentage" | "FixedAmount";
  minOrderAmount: number;
  validFrom: string;
  validTo: string;
  usageLimit: number;
  usedCount: number;
}

// API Types
export interface PlaceOrderRequest {
  userId: string;
  items: Array<{
    bookId: string;
    quantity: number;
  }>;
  shippingAddress: Omit<OrderAddress, "id" | "orderId">;
  couponCode?: string;
}

export interface PlaceOrderResponse {
  order: Order;
  qrCodeUrl: string;
}

export interface User {
  id: string;
  email: string;
  fullName?: string;
  phoneNumber?: string;
}
